presto-main/src/main/java/com/facebook/presto/operator/aggregation/OptimizedTypedSet.java

```diff
@@ -0,0 +1,355 @@
+    }
+
+    public long getRetainedSizeInBytes()
+    {
+        long sizeInBytes = INSTANCE_SIZE + ARRAY_LIST_INSTANCE_SIZE + sizeOf(blockPositionByHash);
+        for (int i = 0; i <= currentBlockIndex; i++) {
+            sizeInBytes += sizeOf(positionsForBlocks.get(i).getPositions());
+        }
+        return sizeInBytes;
+    }
+
+    public void ensureBlocksCapacity(int capacity)
+    {
+        if (blocks == null || blocks.length < capacity) {
+            blocks = Arrays.copyOf(blocks, capacity);
+        }
+    }
+
+    private int getMaskedHash(long rawHash)
+    {
+        return (int) (rawHash & hashMask);
+    }
+
+    /**
+     * Return the position in the hashtable the element at Block position should be inserted.
+     *
+     * @return The position of the hashtable to be inserted if the element does not exist, INVALID_POSITION otherwise.
+     * @hashPosition The position into the hashtable the linear probing starts
+     */
+    private int getInsertPosition(long[] hashtable, int hashPosition, Block block, int position)
+    {
+        while (true) {
+            long blockPosition = hashtable[hashPosition];
+            // Doesn't have this element
+            if (blockPosition == EMPTY_SLOT) {
+                // does not exist
+                return hashPosition;
+            }
+            else {
+                // Already has this element
+                int blockIndex = (int) ((blockPosition & 0xffff_ffff_0000_0000L) >> 32);
+                int positionWithinBlock = (int) (blockPosition & 0xffff_ffff);
+                if (positionEqualsPosition(elementType, blocks[blockIndex], positionWithinBlock, block, position)) {
+                    return INVALID_POSITION;
+                }
+            }
+
+            hashPosition = getMaskedHash(hashPosition + 1);
+        }
+    }
+
+    /**
+     * Add an element to the hash table if it's not already existed.
+     *
+     * @param hashtable The target hash table the element to be inserted into
+     * @param hashPosition The position into the hashtable the linear probing starts
+     * @return true of the element is inserted, false otherwise
+     */
+    private boolean addElement(long[] hashtable, int hashPosition, Block block, int position)
+    {
+        while (true) {
+            long blockPosition = hashtable[hashPosition];
+
+            // Doesn't have this element
+            if (blockPosition == EMPTY_SLOT) {
+                hashtable[hashPosition] = ((long) currentBlockIndex << 32) | position;
+                return true;
+            }
+            else {
+                // Already has this element
+                int blockIndex = (int) ((blockPosition & 0xffff_ffff_0000_0000L) >> 32);
+                int positionWithinBlock = (int) (blockPosition & 0xffff_ffff);
+                if (positionEqualsPosition(elementType, blocks[blockIndex], positionWithinBlock, block, position)) {
+                    return false;
+                }
```

Comment:
the else part can be extracted into a function since ```getInsertPosition``` is also doing the same thing more or less.

Reply:
@sujay-jain I understand they share lots of similarities but there are some subtle differences. `addElement` returns boolean while `getInsertPosition` returns int. `addElement` does insert the element into the hashtable but `getInsertPosition` doesn't. The reason having both was to avoid expensive hash value calculation when there're multiple hash tables, so that calculating of the hash value on the first hash table can be reused by the second. I can change the `addElement` to return an int, but then the callsite would become something like this:
```
if (addElement(newBlockPositionByHash, hash, block, i) == INVALID_POSITION) {
                        positions[positionsIndex++] = i;
                    }
```
This again looks not as easy to understand than returning a boolean. It is common in multiple standard libraries that a function to add a new element into a set to return boolean than an int. What do you think?